dora/services/migration_utils.py

```diff
@@ -0,0 +1,193 @@
+        subcategories_to_add_ids.append(subcategory.id)
+
+    for service in Service.objects.filter(subcategories=if_subcategory):
+        sub_categories = extract_subcategories(service)
+
+        if if_subcategory_value in sub_categories:
+            # Ajout des thématiques
+            new_categories_ids = [s["id"] for s in service.categories.values()]
+            for id in categories_to_add_ids:
+                new_categories_ids.append(id)
+
+            service.categories.set(list(set(new_categories_ids)))
+
+            # Ajout des besoins
+            new_subcategories_ids = [s["id"] for s in service.subcategories.values()]
+            for id in subcategories_to_add_ids:
+                new_subcategories_ids.append(id)
+
+            service.subcategories.set(list(set(new_subcategories_ids)))
+
+            # Sauvegarde
+            Service.objects.filter(pk=service.pk).first().subcategories.set(
+                list(set(new_subcategories_ids))
+            )
+            Service.objects.filter(pk=service.pk).first().categories.set(
+                list(set(new_categories_ids))
+            )
+
+
+def create_category(ServiceCategory, value, label):
+    ServiceCategory.objects.create(value=value, label=label)
+
+
+def get_subcategory_by_value(ServiceSubCategory, value):
+    return ServiceSubCategory.objects.filter(value=value).first()
+
+
+def get_category_by_value(ServiceCategory, value):
+    return ServiceCategory.objects.filter(value=value).first()
+
+
+def create_subcategory(ServiceSubCategory, value, label):
+    ServiceSubCategory.objects.create(value=value, label=label)
+
+
+def update_subcategory_value_and_label(
+    ServiceSubCategory, old_value, new_value, new_label
+):
+    old_subcategory = get_subcategory_by_value(ServiceSubCategory, old_value)
+    if old_subcategory is None:
+        raise ValidationError(f"Aucun besoin trouvé avec la value: '{old_value}'")
+
+    new_subcategory = get_subcategory_by_value(ServiceSubCategory, new_value)
+    if new_subcategory is not None:
+        raise ValidationError(f"La value '{new_value}' est déjà utilisé")
+
+    old_subcategory.value = new_value
+    old_subcategory.label = new_label
+    old_subcategory.save()
+
+
+def replace_subcategory(ServiceSubCategory, Service, from_value, to_value):
+    """
+    Met à jour tous les services en :
+        - retirant le besoin `from_value`
+        - ajoutant le besoin `to_value`
+    """
+    from_subcategory = get_subcategory_by_value(ServiceSubCategory, from_value)
+    if from_subcategory is None:
+        raise ValidationError(f"Aucun besoin trouvé avec la value: '{from_value}'")
+
+    to_subcategory = get_subcategory_by_value(ServiceSubCategory, to_value)
+    if to_subcategory is None:
+        raise ValidationError(f"Aucun besoin trouvé avec la value: '{to_value}'")
+
+    for service in Service.objects.all():
+        sub_categories = extract_subcategories(service)
+
+        if from_value in sub_categories:
```

Comment:
pas besoin de passer par values, ni d'extraire l'id 
`new_subcategories = [s for s in service.subcategories.all() if s.value != from_value]`

Mais ça aurait été encore plus simple de faire un remove suivi d'un add.
Next time!

Reply:
J'ai rajouté un TODO